pandas/core/index.py

```diff
@@ -3818,6 +3842,549 @@ def _wrap_joined_index(self, joined, other):
+    copy : bool, default False
+        Make a copy of input if its a RangeIndex
+
+    Notes
+    -----
+    start can also be a RangeIndex, or a range object
+
+    """
+
+    _typ = 'rangeindex'
+    _engine_type = _index.Int64Engine
+
+    def __new__(cls, start=None, stop=None, step=None, name=None, dtype=None, fastpath=False, copy=False, **kwargs):
+        if fastpath:
+            return cls._simple_new(start, stop, step, name=name)
+
+        # RangeIndex / range
+        if isinstance(start, (RangeIndex, range)):
+
+            if dtype is None or is_int64_dtype(dtype):
+                pass
+            elif is_object_dtype(dtype):
+                # coerce to object
+                return Index(np.array(start), name=name, dtype=dtype, **kwargs)
+            else:
+                raise TypeError('Invalid to pass a non-int64 dtype to RangeIndex')
+
+            if isinstance(start, RangeIndex):
+
+                if not copy:
+                    return start
+                if name is None:
+                    name = getattr(start,'name',None)
+                start, stop, step = start.start, start.stop, start.step
+
+            else:
+
+                # seems we only have indexing ops to infer
+                # rather than direct accessors
+                try:
+                    step = start[1]-start[0]
+                except IndexError:
+                    step = 1
+                stop = start[-1] + step
+                start = start[0]
+
+        # cheap check for array input
+        if len(kwargs) > 0:
+            return cls._data_passthrough(start, stop, step, name, fastpath, **kwargs)
+
+        # require dtype to be None or int64
+        if not (dtype is None or is_int64_dtype(dtype)):
+            raise TypeError('Invalid to pass a non-int64 dtype to RangeIndex')
+
+        new_start, new_stop, new_step = None, None, None
+
+        # sort the arguments depending on which are provided
+        if start is None:
+            start = 0
+        if step is None:
+            new_step = 1
+        if stop is None:
+            new_stop = start
+            new_start = 0
+
+        try:
+            new_start = start if new_start is None else new_start
+            new_stop = stop if new_stop is None else new_stop
+            new_step = step if new_step is None else new_step
+            new_start = cls._ensure_int(new_start)
+            new_stop = cls._ensure_int(new_stop)
+            new_step = cls._ensure_int(new_step)
+            if new_step == 0:
+                raise ValueError("Step must not be zero")
+            return cls._simple_new(new_start, new_stop, new_step, name)
+        except TypeError:
+            return cls._data_passthrough(start, stop, step, name, fastpath=False, **kwargs)
+
+    @classmethod
+    def _simple_new(cls, start, stop=None, step=None, name=None, dtype=None, **kwargs):
+        result = object.__new__(cls)
+
+        # accept None -> (0,0,1)
```

Comment:
this turns out is for empty reindexing, we do something like: ``index._simple_new(None, **attrs)``. So this is necessary :< (but everything else is fixed up)